plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs

```diff
@@ -106,25 +110,38 @@ instance Pretty RestrictingSt where
 -- | For execution, to avoid overruns.
 restricting :: forall uni fun . (PrettyUni uni fun) => ExRestrictingBudget -> ExBudgetMode RestrictingSt uni fun
 restricting (ExRestrictingBudget (ExBudget cpuInit memInit)) = ExBudgetMode $ do
-    -- Using two separate 'STRef's instead of a single one for efficiency reasons.
-    -- Gave us a ~1% speedup the time this idea was implemented.
-    cpuRef <- newSTRef cpuInit
-    memRef <- newSTRef memInit
-    let spend _ (ExBudget cpuToSpend memToSpend) = do
-            cpuLeft <- readSTRef cpuRef
-            memLeft <- readSTRef memRef
+    -- We keep the counters in a PrimArray. This is better than an STRef since it stores its contents unboxed.
+    --
+    -- If we don't specify the element type then GHC has difficulty inferring it, but it's
+    -- annoying to specify the monad, since it refers to the 's' which is not in scope.
+    -- So we use a partial type signature.
+    (ref :: MutablePrimArray _ SatInt) <- newPrimArray 2
```

Comment:
I'd write that as `ref <- newPrimArray @_ @SatInt` but whatever.

Reply:
Ah yeah, I always forget about `@_`